File: spiral_numbers/spiral_numbers.py

# ---------------------------------------------------------
def spiral_numbers_1(N):
    nums = [[0 for i in range(N)] for i in range(N)]
    c1 = r1 = 0
    c2 = r2 = N-1
    val = 1
    while c1 < c2 and r1 < r2:
        for i in range(c1, c2 + 1):
            nums[r1][i] = val
            val += 1
        r1 += 1
        for i in range(r1, r2 + 1):
            nums[i][c2] = val
            val += 1
        c2 -= 1
        for i in range(c2, c1-1, -1):
            nums[r2][i] = val
            val += 1
        r2 -= 1
        for i in range(r2, r1-1, -1):
            nums[i][c1] = val
            val += 1
        c1 += 1
    return nums

def pretty_print(nums):
    max_num = max([max(row) for row in nums])
    max_spaces = len(str(max_num))

    for row in nums:
        for i in row:
            len_i = len(str(i))
            end_spaces = " " * (max_spaces - len_i)
            print(i, end=end_spaces + " ")
        print()

# ---------------------------------------------------------
N = 4
result = spiral_numbers_1(N)
pretty_print(result)

File: spiral_numbers/spiral_numbers.cpp
// g++ -std=c++17 spiral_numbers.cpp

#include <iostream>    // std::cout, std::endl
#include <array>       // std::array
#include <string>      // std::to_string

void pretty_print(int* M, int N) {
    // Determine max string length that needs to be accounted for
    int max_num = 0;
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            int curr_num = *(M + i*N + j);
            int len_i = std::to_string(curr_num).length();
            std::cout << len_i << " ";
            if (curr_num > max_num) {
                max_num = curr_num;
            }
        }
        std::cout << std::endl;
    }
    int max_spaces = std::to_string(max_num).length();
    std::cout << max_spaces << std::endl;
    std::cout << std::endl;

    // Print
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            int curr_num = *(M + i*N + j);
            int len_i = std::to_string(curr_num).length();
            std::string end_spaces = std::string(max_spaces - len_i, ' ');
            std::cout << curr_num << end_spaces << " ";
        }
        std::cout << std::endl;
    }
}

int main() {
    // Set up
    int N = 4;
    int M[N][N];
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            M[i][j] = 0;
        }
    }

    // Fill
    int curr_r = 0;
    int curr_c = 0;
    int dir_r = 0;
    int dir_c = +1;
    M[curr_r][curr_c] = 1;
    int val = 2;
    while (val <= N*N) {
        if (curr_r + dir_r < N && curr_c + dir_c < N &&
            curr_r + dir_r >= 0 && curr_r + dir_c >= 0 &&
            M[curr_r + dir_r][curr_c + dir_c] == 0) {
            M[curr_r + dir_r][curr_c + dir_c] = val;
            val += 1;
            curr_r += dir_r;
            curr_c += dir_c;
        } else {
            if (dir_r == 0 && dir_c == 1) {
                dir_r = 1;
                dir_c = 0;
            }
            else if (dir_r == 1 && dir_c == 0) {
                dir_r = 0;
                dir_c = -1;
            }
            else if (dir_r == 0 && dir_c == -1) {
                dir_r = -1;
                dir_c = 0;
            }
            else if (dir_r == -1 && dir_c == 0) {
                dir_r = 0;
                dir_c = 1;
            }
        }
    }
    pretty_print(*M, N);
}
